### PR TITLE
WIP - apiextensions: low-level converter wiring – towards versioning

### PR DESCRIPTION
**Pull request by sttts**
- add crdConversionRESTOptionsGetter to wire in custom conversions in storage path
- wire through custom conversions in serializers for request de- and encoding.

### CODE REVIEW

File: staging/src/k8s.io/apimachinery/pkg/runtime/serializer/versioning/versioning.go [hunk 1]
```diff
@@ -170,17 +171,20 @@ func (c *codec) Encode(obj runtime.Object, w io.Writer) error {
...
-		if len(objGVK.Version) == 0 {
-			return c.encoder.Encode(obj, w)
-		}
-		targetGVK, ok := c.encodeVersion.KindForGroupVersionKinds([]schema.GroupVersionKind{objGVK})
-		if !ok {
-			return runtime.NewNotRegisteredGVKErrForTarget(objGVK, c.encodeVersion)
-		}
-		if targetGVK == objGVK {
-			return c.encoder.Encode(obj, w)
+		// Only avoid conversion if unstructured is not a list, giving the converter a chance to convert list items.
```
Comment (deads2k) [inline staging/src/k8s.io/apimachinery/pkg/runtime/serializer/versioning/versioning.go]: When did you find a list that asserted the "correct" group/version, but actually contained types in the wrong group/version?

File: staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/etcd_test.go [hunk 2]
```diff
@@ -59,7 +59,7 @@ func newStorage(t *testing.T) (customresource.CustomResourceStorage, *etcdtestin
 		&metav1.DeleteOptions{},
 	)
 
-	typer := apiserver.UnstructuredObjectTyper{
+	typer := apiserver.unstructuredObjectTyper{
```
Comment (deads2k) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/etcd_test.go]: This doesn't compile, does it?

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go [hunk 3]
```diff
@@ -640,3 +683,29 @@ func (in crdStorageMap) clone() crdStorageMap {
...
+
+func (t crdConversionRESTOptionsGetter) GetRESTOptions(resource schema.GroupResource) (generic.RESTOptions, error) {
+	ret, err := t.delegate.GetRESTOptions(resource)
+	if err == nil {
+		ret.StorageConfig.Codec = versioning.NewCodec(
+			ret.StorageConfig.Codec,
+			ret.StorageConfig.Codec,
+			t.converter,
+			&unstructuredCreator{},
+			discovery.NewUnstructuredObjectTyper(nil),
```
Comment (deads2k) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go]: reminder to myself, I promised to move this to the extensions apiserver because it's the only consumer and it's misnamed.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go [hunk 4]
```diff
@@ -640,3 +683,29 @@ func (in crdStorageMap) clone() crdStorageMap {
 	}
 	return out
 }
+
+// crdConversionRESTOptionsGetter overrides the codec with one using the custom converter
+// and custom encoder and decoder version.
+type crdConversionRESTOptionsGetter struct {
+	delegate       generic.RESTOptionsGetter
+	converter      runtime.ObjectConvertor
```
Comment (deads2k) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go]: please doc granularity.  All CRDs or one group? or something else?

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go [hunk 5]
```diff
@@ -535,9 +548,37 @@ func (c crdObjectConverter) ConvertFieldLabel(version, kind, label, value string
...
+
+	if unstructList, ok := in.(*unstructured.UnstructuredList); ok {
+		gvk, ok := target.KindForGroupVersionKinds([]schema.GroupVersionKind{kind})
+		if !ok {
+			return nil, fmt.Errorf("%v is unstructured and is not suitable for converting to %q", kind, target)
+		}
+		in.GetObjectKind().SetGroupVersionKind(gvk)
+
+		for i := range unstructList.Items {
+			out, err := c.unstructuredDelegate.ConvertToVersion(&unstructList.Items[i], target)
```
Comment (deads2k) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go]: Seems weird to directly delegate instead of recursing.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go [hunk 6]
```diff
@@ -517,10 +522,18 @@ func (r *crdHandler) getOrCreateServingInfoFor(crd *apiextensions.CustomResource
...
 
-// crdObjectConverter is a converter that supports field selectors for CRDs.
+// crdObjectConverter wraps an unstructured object converter adding:
+// - support for field selectors
+// - support for UnstructuredList, applying the delegate to each item.
 type crdObjectConverter struct {
-	unstructured.UnstructuredObjectConverter
-	clusterScoped bool
+	unstructuredDelegate unstructured.UnstructuredObjectConverter
+	clusterScoped        bool
```
Comment (deads2k) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go]: cruft?

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go [hunk 7]
```diff
@@ -382,12 +382,17 @@ func (r *crdHandler) getOrCreateServingInfoFor(crd *apiextensions.CustomResource
...
 	parameterCodec := runtime.NewParameterCodec(parameterScheme)
 
+	clusterScoped := crd.Spec.Scope == apiextensions.ClusterScoped
 	kind := schema.GroupVersionKind{Group: crd.Spec.Group, Version: crd.Spec.Version, Kind: crd.Status.AcceptedNames.Kind}
-	typer := UnstructuredObjectTyper{
-		Delegate:          parameterScheme,
-		UnstructuredTyper: discovery.NewUnstructuredObjectTyper(),
+	typer := unstructuredObjectTyper{
+		delegate:          parameterScheme,
+		unstructuredTyper: discovery.NewUnstructuredObjectTyper(),
```
Comment (deads2k) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go]: This is on its way to being a lot more opinionated.  How about starting a `scheme` package or some such?

**Comment by deads2k** [comment]
I think I see where you're going with it.  It makes sense.

Can you summarized the intent in the description?  It appears to me that you're working towards building a mutable object converter that is going to be used inside of the "normal" codec flow and will thus allow you to take a lock and change the converter as needed.

**Comment by deads2k** [comment]
@kubernetes/sig-api-machinery-misc @lavalamp

File: staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/etcd_test.go (hunk 2 above)
Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/etcd_test.go]: No, it doesn't. We shouldn't leak from the apiserver into the registry in the first place though.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go (hunk 5 above)
Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go]: We can't recurse because the underlying converter will be different (think of declarative conversion some day). Hence, we make this composition, so the generic crd converter can call the delegate converter doing the actual work on non-lists.

**Comment by lavalamp** [comment]
Can this be rebased on top of #63830? Or is it trying to be an alternative approach?

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go [hunk 8]
```diff
@@ -535,9 +548,37 @@ func (c crdObjectConverter) ConvertFieldLabel(version, kind, label, value string
 	}
 }
 
+func (c crdObjectConverter) ConvertToVersion(in runtime.Object, target runtime.GroupVersioner) (runtime.Object, error) {
+	kind := in.GetObjectKind().GroupVersionKind()
+	if kind.Empty() {
+		return in, nil
```
[the remark this answers is not in the document]
Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go]: here the analysis and a possible fix: https://github.com/mbohlool/kubernetes/pull/5

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go (hunk 3 above)
Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go]: We might want to move all of these into a scheme package here.

**Comment by deads2k** [comment]
> Can this be rebased on top of #63830? Or is it trying to be an alternative approach?

My read is that this is doing the underlying wiring that is logically needed in #63830 and @sttts is helping actually do it instead of simply pushing a many comments into the pull.

**Comment by sttts** [comment]
> My read is that this is doing the underlying wiring that is logically needed in #63830

Yes, this is 90% the same code as in #63830, minus adding the multiple versions. The goal was to iterate quicker with reviews than mixing lower level plumbing with API discussion in a big API implementing PR. The later takes a magnitude longer for reviewed and to get it merged. Also having some soak time for those low-level changes at the heart of CRDs wouldn't harm.

These changes here are also blocking work on pruning and defaulting, which puts it on the critical path towards code-freeze (and with pruning being prerequisite for GA also for promotion of CRDs).

I don't feel strongly about which PR we finally merge (code is mostly the same), as long as we get it in this week. So if we put enough priority on #63830 both for review and for addressing comments, that's fine as well.
